### PR TITLE
feat: add rendezvous API [DET-5428]

### PR DESCRIPTION
**Pull request by stoksc**
## Description
This change adds a rendezvous API for trials in a post-push-arch world. The API is designed to block until all containers are connected and their rendezvous info has arrived from container started message. Once these conditions are met, rendezvous is pushed to listening watchers.

## Test Plan
- [x] kinda hard but, call API and make sure it performs as expected.

## Commentary (optional)
Went with chan's as opposed to actors again. The number of messages is bounded and the amount of code is significantly less than an actor, I think it's much clearer, tbh.

## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: proto/src/determined/api/v1/trial.proto [hunk 1]
```diff
@@ -376,3 +376,21 @@ message ReportTrialCheckpointMetadataRequest {
...
+  int32 trial_id = 1;
+  // The id of the container. Used to verify all containers are connected.
+  string container_id = 2;
+  // The rendezvous info.
+  determined.trial.v1.RendezvousInfo rendezvous_info = 3;
+}
+// Response to GetTrialRendezvousInfoRequest
+message GetTrialRendezvousInfoResponse {
+  // The rendezvous addresses.
+  repeated string addresses = 1;
```
Comment (shiyuann) [inline proto/src/determined/api/v1/trial.proto:395]: This should use `determined.trial.v1.RendezvousInfo` struct?

File: master/internal/trial.go [hunk 2]
```diff
@@ -105,12 +106,23 @@ type (
...
 	}
 
-	// trialWatchPreemption begins watching if the trial has been preempted.
+	// trialWatchPreemptionReq begins watching if the trial has been preempted.
 	// The trial responds to this message with a channel of bools, where sends of true
 	// indicate to preempt and sends of false are used to synchronize (e.g. you want to
 	// block until you receive _something_ but not until the first preemption).
-	trialWatchPreemption   struct{ id uuid.UUID }
-	trialUnwatchPreemption struct{ id uuid.UUID }
+	trialWatchPreemptionReq  struct{ id uuid.UUID }
```
Comment (shiyuann) [inline master/internal/trial.go:113]: Can you name it to be `trialWatchPreemption`?

```suggestion
	trialWatchPreemption  struct{ id uuid.UUID }
```

File: master/internal/trial.go [hunk 3]
```diff
@@ -105,12 +106,23 @@ type (
...
+	trialWatchPreemptionResp struct{ signal <-chan bool }
+	trialUnwatchPreemption   struct{ id uuid.UUID }
+
+	// trialWatchRendezvousInfoReq begins watching for rendezvous info.
+	// When all the containers are ready, the trial will send all the
+	// peer addresses on the channel in the response.
+	trialWatchRendezvousInfoReq  struct{ containerID cproto.ID }
```
Comment (shiyuann) [inline master/internal/trial.go:120]:
```suggestion
	trialWatchRendezvousInfo  struct{ containerID cproto.ID }
```

File: master/internal/trial.go [hunk 4]
```diff
@@ -206,7 +218,9 @@ type (
 		sequencer *trialWorkloadSequencer
 
 		// The following fields tracks the interaction with the resource providers.
-		task        *sproto.AllocateRequest
+		// The existence of task signifies we have asked to be allocated.
```
Comment (shiyuann) [inline master/internal/trial.go:221]:
```suggestion
		// The existence of task signifies the trial has requested to be allocated.
```

File: master/internal/trial.go [hunk 5]
```diff
@@ -206,7 +218,9 @@ type (
...
+		task *sproto.AllocateRequest
+		// The existence of allocations signify we asked to be allocated and were allocated.
```
Comment (shiyuann) [inline master/internal/trial.go:223]:
```suggestion
		// The existence of allocations signifies the trial has been allocated.
```

File: master/internal/trial.go [hunk 6]
```diff
@@ -394,6 +406,59 @@ func (t *trial) Receive(ctx *actor.Context) error {
...
+		return
+	}
+
+	// Register and respond with the watcher channel. Size 2; at most 2 messages
+	// can be sent and we don't want to block or lose them.
+	w := make(chan bool, 2)
+	t.preemptionWatchers[msg.id] = w
+	ctx.Respond(trialWatchPreemptionResp{signal: w})
+
+	if t.PendingGracefulTermination {
```
Comment (shiyuann) [inline master/internal/trial.go:453]: Why did you change this?

**Review by shiyuann — COMMENTED** [review]
I'd like to decouple the new preemption logic entirely from the trial logic or the termination logic. For example, define a go struct that only contains the preemption logic. Besides, we need to have a test for this change.

File: master/internal/trial.go [hunk 7]
```diff
@@ -935,6 +1091,11 @@ func (t *trial) processContainerRunning(
 
 	t.containers[msg.Container.ID] = msg.Container
 	t.containerAddresses[msg.Container.ID] = msg.ContainerStarted.Addresses
+	// TODO(DET-5280): Uncomment this code, nuke below it.
```
Comment (shiyuann) [inline master/internal/trial.go:1094]: I don't think you are linking to the correct ticket number here. Besides, can we just have a global switch in the master to choose between V1 and V2. I don't want the V2 logic to be heavily dependent on V1 logic. It would be bad for writing tests and hard to switch to V2 cleanly.

Comment (stoksc) [reply master/internal/trial.go:1094]: > I don't think you are linking to the correct ticket number here.

I just linked the push arch epic.

>  I don't want the V2 logic to be heavily dependent on V1 logic.

It's not dependent at all, once v2 is on, v1 will not exist at all.

**Comment by stoksc** [comment]
> For example, define a go struct that only contains the preemption logic.

I'm not sure at all what you mean by this, we should just chat offline.

File: master/internal/trial.go (hunk 6 above)
Comment (stoksc) [reply master/internal/trial.go:453]: I felt `len(allocations) == 0` was a more correct check than `trial_state != ACTIVE`.

File: proto/src/determined/api/v1/trial.proto (hunk 1 above)
Comment (stoksc) [reply proto/src/determined/api/v1/trial.proto:395]: Yeah, I had that struct around before I rebased onto rb's trim rendezvous info pr. I suppose I could still use it, just meant to remove though.

File: proto/src/determined/api/v1/trial.proto [hunk 8]
```diff
@@ -376,3 +376,21 @@ message ReportTrialCheckpointMetadataRequest {
...
+}
```
Comment (rb-determined-ai) [inline proto/src/determined/api/v1/trial.proto:396]: I think we need a machine rank in the response too.

**Review by rb-determined-ai — COMMENTED** [review]
Working on the python rendezvous layer, but I hit a few snags.

File: proto/src/determined/api/v1/trial.proto [hunk 9]
```diff
@@ -376,3 +376,21 @@ message ReportTrialCheckpointMetadataRequest {
...
+message GetTrialRendezvousInfoRequest {
+  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
+    json_schema: { required: [ "trial_id", "rendezvous_info" ] }
+  };
+  // The id of the trial.
+  int32 trial_id = 1;
+  // The id of the container. Used to verify all containers are connected.
+  string container_id = 2;
+  // The rendezvous info.
+  determined.trial.v1.RendezvousInfo rendezvous_info = 3;
```
[the remark this answers is not in the document]
Comment (stoksc) [reply proto/src/determined/api/v1/trial.proto:390]: 🤷 i honestly have no idea why that's there.. i dont remember doing that. maybe bad rebase.

File: proto/src/determined/api/v1/trial.proto [hunk 10]
```diff
@@ -376,3 +376,21 @@ message ReportTrialCheckpointMetadataRequest {
 }
 // Response to ReportTrialCheckpointMetadataRequest
 message ReportTrialCheckpointMetadataResponse {}
+
+// Gathers the rendezvous info for the trial.
+message GetTrialRendezvousInfoRequest {
+  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
+    json_schema: { required: [ "trial_id", "rendezvous_info" ] }
```
[the remark this answers is not in the document]
Comment (stoksc) [reply proto/src/determined/api/v1/trial.proto:383]: also fixed, not sure what happened here

File: master/internal/trial.go (hunk 7 above)
Comment (stoksc) [reply master/internal/trial.go]: done

File: proto/src/determined/api/v1/trial.proto (hunk 1 above)
Comment (stoksc) [reply proto/src/determined/api/v1/trial.proto]: done

File: proto/src/determined/api/v1/trial.proto (hunk 8 above)
Comment (stoksc) [reply proto/src/determined/api/v1/trial.proto:397]: done

File: proto/src/determined/api/v1/api.proto [hunk 11]
```diff
@@ -588,6 +588,17 @@ service Determined {
       tags: "Internal"
     };
   }
+  // Gather a trial's rendezvous info. Blocks until all trial containers connect
+  // to gather their rendezvous information and responds to them all at once.
+  rpc GetTrialRendezvousInfo(GetTrialRendezvousInfoRequest)
+      returns (GetTrialRendezvousInfoResponse) {
+    option (google.api.http) = {
+      get: "/api/v1/trials/{trial_id}/rendezvous_info/{container_id}"
```
Comment (stoksc) [inline proto/src/determined/api/v1/api.proto:596]: wdyt @rb-determined-ai ? this or just make it a POST?

File: master/internal/trial.go (hunk 6 above)
Comment (stoksc) [reply master/internal/trial.go:462]: ended up adding cleanup in `t.terminated()` too, so the lifetime of a rendezvous or preemption watcher should only be from one `sproto.ResourcesAllocated` to the `t.terminated(ctx)` call to clean them up. This lifetime will become cleaner in an upcoming refactor.

File: master/internal/trial_test.go [hunk 12]
```diff
@@ -46,9 +46,6 @@ func TestRendezvousInfo(t *testing.T) {
 			{
 				ContainerPort: 1,
 			},
-			{
```
Comment (stoksc) [inline master/internal/trial_test.go:49]: `chore: trim unused parts of rendezvous info (#2381)` trimmed a rendezvous port from the response, but since the master never errored on wrong rendezvous (it just dropped them and moved on) this extra rendezvous resulted in incorrect rendezvous info and an error log but no test failure. I altered this behavior (intentionally) so that the trial would fail too, which resulted in me noticing this. i restored the old behavior but also fixed the test.

File: master/internal/trial.go [hunk 13]
```diff
@@ -1230,3 +1305,75 @@ func (t *trial) tellWithSnapshot(
...
+	watchPreemption   struct{ id uuid.UUID }
+	preemptionWatcher struct{ C <-chan struct{} }
+	unwatchPreemption struct{ id uuid.UUID }
+
+	// preemption represents the preemption status of a task. A task is assumed to be preempted
+	// exactly one time. The object is "nil safe" - it'll gracefully handle calls on a nil
+	// preemption. This is nice until we move to trial has many task actors / generic task actor,
+	// where the lifetime of a "preemption" is equivalent to the lifetime of task and they can be
+	// initialized together.
+	preemption struct {
```
Comment (stoksc) [inline master/internal/trial.go:1323]: long term, when push arch is done, i plan to do some refactoring. i will move this somewhere. i was thinking it could go into internal/task or internal/job or something. maybe pkg/task could be renamed pkg/taskspec

File: master/internal/trial_test.go [hunk 14]
```diff
@@ -188,3 +182,61 @@ func TestRendezvousInfo(t *testing.T) {
...
+	// "task" is allocated.
+	p = newPreemption()
+
+	// real watcher connects
+	id = uuid.New()
+	w, err := p.watch(watchPreemption{id: id})
+	assert.NilError(t, err)
+
+	// should immediately receive initial status.
+	select {
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/trial_test.go:209]: why would I use a goroutine for this, what do you mean?

File: master/internal/trial.go [hunk 15]
```diff
@@ -829,50 +875,90 @@ func (t *trial) killAndRemoveSocket(ctx *actor.Context, id cproto.ID) {
...
+			w := t.rendezvousWatchers[c.ID]
+			if err != nil {
+				w <- err
+			} else {
+				w <- &trialv1.RendezvousInfo{
+					Addresses: raddrs,
+					Rank:      int32(t.containerRanks[c.ID]),
+				}
+			}
+			close(w)
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/trial.go:917]: what do you mean `closeRendezvous`? this is just a channel and the `close` builtin.

File: master/internal/trial.go [hunk 16]
```diff
@@ -105,12 +112,15 @@ type (
...
-	// block until you receive _something_ but not until the first preemption).
-	trialWatchPreemption   struct{ id uuid.UUID }
-	trialUnwatchPreemption struct{ id uuid.UUID }
+	// trialWatchRendezvousInfoReq begins watching for rendezvous info.
+	// When all the containers are ready, the trial will send all the
+	// peer addresses on the channel in the response.
+	watchRendezvousInfo struct{ containerID cproto.ID }
+	rendezvousWatcher   struct {
+		// interface{} is *trialv1.RendezvousInfo or error.
+		C <-chan interface{}
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/trial.go:121]: sure, though i'm going to rewrite this as soon as go 1.18 is out, fwiw.

File: master/internal/trial.go [hunk 17]
```diff
@@ -225,7 +237,11 @@ type (
 		privateKey     []byte
 		publicKey      []byte
 
-		preemptionWatchers map[uuid.UUID]chan<- bool
+		// preemption represents the preemption state of the current allocated task.
+		// If there is no current task, or it is unallocated, it is nil.
+		preemption *preemption
+		// Map of container ID to watcher ID a rendezvous info listener.
+		rendezvousWatchers map[cproto.ID]chan<- interface{}
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/trial.go:244]: i did think about this, but there were no methods i planned on writing that were anything more than thin wrappers around well known channel primitives.

File: master/internal/trial.go [hunk 18]
```diff
@@ -394,6 +412,35 @@ func (t *trial) Receive(ctx *actor.Context) error {
...
+	}
+
+	// Channel is size 1 since rendezvous info will only ever be sent once.
+	w := make(chan interface{}, 1)
+	t.rendezvousWatchers[msg.containerID] = w
+
+	t.lastContainerConnectedTime = time.Now()
+	if !t.allReady() {
+		actors.NotifyAfter(ctx, allReadyTimeoutPeriod, allReadyTimeout{runID: t.RunID})
+		ctx.Log().Info("found not all containers are connected")
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/trial.go:436]: sure, but without the period.

**Comment by stoksc** [comment]
@shiyuann had a lot of questions so i went ahead and reassigned before i turn around.

File: master/internal/trial.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/trial.go:917]: I see what you mean.. `closeRendezvous` tells all watchers still watching that the task was terminated and they missed rendezvous. This sends the rendezvouz and just closes the chans. maybe closeRendezvous is more aptly called terminateRendezvousWatchers

File: master/internal/trial.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/trial.go:439]:
```suggestion
		ctx.Log().Debug("not sending rendezvous information because not all trial containers are connected")
```

**Review by shiyuann — APPROVED** [review]

File: master/internal/trial_test.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (shiyuann) [reply master/internal/trial_test.go:209]: Since all the cases are enumerated here, concurrent testing might not be necessary

File: proto/src/determined/api/v1/api.proto (hunk 11 above)
Comment (rb-determined-ai) [reply proto/src/determined/api/v1/api.proto:597]: POST seems more correct to me, since it is not cacheable

File: proto/src/determined/api/v1/api.proto [hunk 19]
```diff
@@ -498,9 +498,10 @@ service Determined {
   // sent immediately, for synchronization purposes. If it is to preempt, that
```
Comment (rb-determined-ai) [inline proto/src/determined/api/v1/api.proto:498]: the rest of this comment needs updating
